File: src/java/foundgine-providers/src/main/java/com/foundgine/providers/storage/sql/query/SqlAuthorizationWriter.java
package com.foundgine.providers.storage.sql.query;

import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.metadata.*;
import com.foundgine.providers.storage.sql.SqlCompiler;
import java.math.BigDecimal;
import java.util.Collection;

/** Lowers the AOT authorization predicate IR into parameterized SQL. */
public final class SqlAuthorizationWriter {
    private SqlAuthorizationWriter() { }
    public static String write(AuthorizationPredicate predicate, EntityMetadata resource, String resourceAlias, Collection<SqlParameterBinding> parameters) {
        if (predicate == null || resource == null || parameters == null) throw new NullPointerException();
        int[] counter = {0}; return writeNode(predicate, resource, resourceAlias, parameters, counter);
    }
    private static String writeNode(AuthorizationPredicate node, EntityMetadata resource, String alias, Collection<SqlParameterBinding> parameters, int[] counter) {
        return switch (node.kind()) {
            case CONTEXT_PARAMETER -> throw new IllegalStateException("A context parameter must be followed by member access.");
            case RESOURCE_PARAMETER -> alias;
            case PARAMETER -> throw new UnsupportedOperationException("Untyped authorization parameters are not supported by the SQL provider.");
            case MEMBER_ACCESS -> writeMember(node, resource, alias, parameters, counter);
            case CONSTANT -> addConstant(node.value(), parameters, counter);
            case EQUAL -> binary(node, "=", resource, alias, parameters, counter);
            case NOT_EQUAL -> binary(node, "<>", resource, alias, parameters, counter);
            case AND -> binary(node, "AND", resource, alias, parameters, counter);
            case OR -> binary(node, "OR", resource, alias, parameters, counter);
            case NOT -> "NOT (" + required(node.left(), resource, alias, parameters, counter) + ")";
        };
    }
    private static String writeMember(AuthorizationPredicate node, EntityMetadata resource, String alias, Collection<SqlParameterBinding> parameters, int[] counter) {
        AuthorizationPredicate target = node.left(); if (target == null) throw new IllegalStateException("Member access has no target.");
        if (target.kind() == AuthorizationPredicateKind.RESOURCE_PARAMETER) {
            String name = node.name(); if (name == null) throw new IllegalStateException("Resource member has no name.");
            FieldMetadata field = resource.effectiveFields().stream().filter(x -> x.name().equals(name)).findFirst().orElse(null);
            if (field == null || field.column() == null) throw new IllegalArgumentException("Authorization resource member '" + resource.name() + "." + name + "' has no storage column mapping.");
            ColumnMetadata column = resource.columns().stream().filter(x -> x.id().equals(field.column().columnId())).findFirst()
                    .orElseThrow(() -> new IllegalArgumentException("Authorization resource member '" + resource.name() + "." + name + "' references a missing column."));
            return SqlCompiler.quoteIdentifier(alias) + "." + SqlCompiler.quoteIdentifier(column.effectiveStorageName());
        }
        if (target.kind() == AuthorizationPredicateKind.CONTEXT_PARAMETER) {
            String contextName = target.name(); if (contextName == null) throw new IllegalStateException("Context parameter has no name.");
            String member = node.name(); if (member == null) throw new IllegalStateException("Context member has no name.");
            String name = "auth" + counter[0]++; parameters.add(new SqlParameterBinding(name, null, null, contextName + "." + member, null)); return "@" + name;
        }
        throw new UnsupportedOperationException("Only direct resource and context member access is supported by the SQL authorization provider.");
    }
    private static String binary(AuthorizationPredicate node, String op, EntityMetadata resource, String alias, Collection<SqlParameterBinding> parameters, int[] counter) {
        return "(" + required(node.left(), resource, alias, parameters, counter) + " " + op + " " + required(node.right(), resource, alias, parameters, counter) + ")";
    }
    private static String addConstant(String value, Collection<SqlParameterBinding> parameters, int[] counter) { String name="auth"+counter[0]++; parameters.add(new SqlParameterBinding(name, parseConstant(value))); return "@"+name; }
    private static Object parseConstant(String value) {
        if (value == null || value.equals("null")) return null;
        if (value.equals("true")) return true; if(value.equals("false")) return false;
        try { return Integer.parseInt(value); } catch(Exception ignored) {}
        try { return Long.parseLong(value); } catch(Exception ignored) {}
        try { return new BigDecimal(value); } catch(Exception ignored) {}
        return value;
    }
    private static String required(AuthorizationPredicate node, EntityMetadata resource, String alias, Collection<SqlParameterBinding> parameters, int[] counter) { if(node==null) throw new IllegalStateException("Authorization predicate node is incomplete."); return writeNode(node,resource,alias,parameters,counter); }
}

File: src/csharp/Foundgine.Providers/Storage/Sql/Query/SqlAuthorizationWriter.cs
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Metadata;

namespace Foundgine.Providers.Storage.Sql.Query;

/// <summary>
/// Lowers the small AOT authorization predicate IR into provider-neutral SQL
/// fragments. Context values remain parameterized and are resolved only at
/// execution time.
/// </summary>
internal static class SqlAuthorizationWriter
{
    public static string Write(
        AuthorizationPredicate predicate,
        EntityMetadata resource,
        string resourceAlias,
        ICollection<SqlParameterBinding> parameters)
    {
        ArgumentNullException.ThrowIfNull(predicate);
        ArgumentNullException.ThrowIfNull(resource);
        ArgumentNullException.ThrowIfNull(parameters);

        var counter = 0;
        return WriteNode(predicate, resource, resourceAlias, parameters, ref counter);
    }

    private static string WriteNode(
        AuthorizationPredicate node,
        EntityMetadata resource,
        string resourceAlias,
        ICollection<SqlParameterBinding> parameters,
        ref int counter) => node.Kind switch
    {
        AuthorizationPredicateKind.ContextParameter =>
            throw new InvalidOperationException("A context parameter must be followed by member access."),
        AuthorizationPredicateKind.ResourceParameter =>
            resourceAlias,
        AuthorizationPredicateKind.Parameter =>
            throw new NotSupportedException("Untyped authorization parameters are not supported by the SQL provider."),
        AuthorizationPredicateKind.MemberAccess => WriteMember(node, resource, resourceAlias, parameters, ref counter),
        AuthorizationPredicateKind.Constant => AddConstant(node.Value, parameters, ref counter),
        AuthorizationPredicateKind.Equal => Binary(node, "=", resource, resourceAlias, parameters, ref counter),
        AuthorizationPredicateKind.NotEqual => Binary(node, "<>", resource, resourceAlias, parameters, ref counter),
        AuthorizationPredicateKind.And => Binary(node, "AND", resource, resourceAlias, parameters, ref counter),
        AuthorizationPredicateKind.Or => Binary(node, "OR", resource, resourceAlias, parameters, ref counter),
        AuthorizationPredicateKind.Not =>
            $"NOT ({WriteRequired(node.Left, resource, resourceAlias, parameters, ref counter)})",
        _ => throw new NotSupportedException(
            $"Authorization predicate '{node.Kind}' is not supported by the SQL provider.")
    };

    private static string WriteMember(
        AuthorizationPredicate node,
        EntityMetadata resource,
        string resourceAlias,
        ICollection<SqlParameterBinding> parameters,
        ref int counter)
    {
        var target = node.Left ?? throw new InvalidOperationException("Member access has no target.");
        if (target.Kind == AuthorizationPredicateKind.ResourceParameter)
        {
            var name = node.Name ?? throw new InvalidOperationException("Resource member has no name.");
            var field = resource.EffectiveFields.FirstOrDefault(x => x.Name == name);
            if (field?.Column is null)
                throw new InvalidOperationException(
                    $"Authorization resource member '{resource.Name}.{name}' has no storage column mapping.");

            var column = resource.Columns.FirstOrDefault(x => x.Id == field.Column.ColumnId)
                         ?? throw new InvalidOperationException(
                             $"Authorization resource member '{resource.Name}.{name}' references a missing column.");
            return
                $"{SqlCompiler.QuoteIdentifier(resourceAlias)}.{SqlCompiler.QuoteIdentifier(column.EffectiveStorageName)}";
        }

        if (target.Kind == AuthorizationPredicateKind.ContextParameter)
        {
            var contextName = target.Name ?? throw new InvalidOperationException("Context parameter has no name.");
            var member = node.Name ?? throw new InvalidOperationException("Context member has no name.");
            var path = contextName + "." + member;
            var name = "auth" + counter++;
            parameters.Add(new SqlParameterBinding(name, null, ContextPath: path));
            return "@" + name;
        }

        throw new NotSupportedException(
            "Only direct resource and context member access is supported by the SQL authorization provider.");
    }

    private static string Binary(
        AuthorizationPredicate node,
        string op,
        EntityMetadata resource,
        string resourceAlias,
        ICollection<SqlParameterBinding> parameters,
        ref int counter)
    {
        var left = WriteRequired(node.Left, resource, resourceAlias, parameters, ref counter);
        var right = WriteRequired(node.Right, resource, resourceAlias, parameters, ref counter);
        return $"({left} {op} {right})";
    }

    private static string AddConstant(string? value, ICollection<SqlParameterBinding> parameters, ref int counter)
    {
        var name = "auth" + counter++;
        parameters.Add(new SqlParameterBinding(name, ParseConstant(value)));
        return "@" + name;
    }

    private static object? ParseConstant(string? value) => value switch
    {
        null => null,
        "null" => null,
        "true" => true,
        "false" => false,
        _ when int.TryParse(value, out var i) => i,
        _ when long.TryParse(value, out var l) => l,
        _ when decimal.TryParse(value, out var d) => d,
        _ => value
    };

    private static string WriteRequired(
        AuthorizationPredicate? node,
        EntityMetadata resource,
        string resourceAlias,
        ICollection<SqlParameterBinding> parameters,
        ref int counter) =>
        node is null
            ? throw new InvalidOperationException("Authorization predicate node is incomplete.")
            : WriteNode(node, resource, resourceAlias, parameters, ref counter);
}